x-pack/plugins/security/server/lib/privileges/privilege_action_registry.js

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { buildPrivilegeMap } from './privileges';
+import { getClient } from '../../../../../server/lib/get_client_shield';
+import { equivalentPrivileges } from './equivalent_privileges';
+
+export async function registerPrivilegesWithCluster(server) {
+  const config = server.config();
+  const kibanaVersion = config.get('pkg.version');
+  const application = config.get('xpack.security.rbac.application');
+  const savedObjectTypes = server.savedObjects.types;
+
+  const expectedPrivileges = {
+    [application]: buildPrivilegeMap(savedObjectTypes, application, kibanaVersion)
+  };
+
+  server.log(['security', 'debug'], `Registering Kibana Privileges with Elasticsearch for ${application}`);
+
+  const callCluster = getClient(server).callWithInternalUser;
+
+  try {
+    // we only want to post the privileges when they're going to change as Elasticsearch has
+    // to clear the role cache to get these changes reflected in the _has_privileges API
+    const existingPrivileges = await callCluster(`shield.getPrivilege`, { privilege: application });
+    if (equivalentPrivileges(existingPrivileges, expectedPrivileges)) {
```

Comment:
I believe that Spencer and I briefly chatted about this over Zoom.

Kibana plugins can also potentially influence the Kibana application privileges, currently only by adding saved object types. There's potential for two different instances of Kibana to have two different sets of plugins installed, causing them to each insert a different set of privileges. 

When Spencer and I were discussing this situation, I suggested ensuring the privileges we were expecting were a "subset" of the ones already persisted to Elasticsearch, and if that was the case not overwriting the application privileges. This was meant to solve the situation where a plugin was installed on one instance of Kibana but not another. After making this suggestion, I realized this doesn't solve the situation where we have two different plugins installed in two different instances of Kibana, and a number of other permutations of differing plugins, so I don't think it's a solution we'll want to adopt.

Reply:
> Kibana plugins can also potentially influence the Kibana application privileges

They'd do this through the security plugin though, so the security plugin still has complete knowledge of all possible privileges whether they are registered from the security plugin on behalf of our open source features or whether they were registered by a plugin.

> There's potential for two different instances of Kibana to have two different sets of plugins installed

What's the use case for this? Off-hand, this doesn't sound like something we should support.

Reply:
So... right now, we're building the `read` privilege [here](https://github.com/elastic/kibana/blob/security-app-privs/x-pack/plugins/security/server/lib/privileges/privileges.js#L39) based on the types that inferred from the saved object mappings [here](https://github.com/elastic/kibana/blob/security-app-privs/src/server/saved_objects/service/create_saved_objects_service.js#L86) so any plugin that specifies it's own mappings.json will automatically work properly with RBAC.

> What's the use case for this? Off-hand, this doesn't sound like something we should support.

It's not something that we'd recommend for sure, and it's something that I was planning on documenting as "not supported", but it could definitely happen, and when it does each instance of Kibana when it starts up will force it's version of the Kibana Application Privileges into Elasticsearch and put us into a rather inconsistent state based on which one started up last. 

There are additional, but more complicated, solutions that we could explore. One of these that Spencer recommended was to allow each instance of Kibana to "append" it's own actions to the privileges as long as the version action matched.